File: Android/java/app/src/main/java/org/oreilly/nmd/SearchResultsActivity.java
package org.oreilly.nmd;

import android.app.Activity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class SearchResultsActivity extends Activity {

  public static final String QUERY = "QUERY";

  private ProgressBar mProgressBar;
  private RecyclerView mRecyclerView;

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_search);
    mProgressBar = findViewById(R.id.search_progress);
    mRecyclerView = findViewById(R.id.search_results_recyclerview);
    mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
    String query = getIntent().getStringExtra(QUERY);
    new Thread(() -> fetchResults("magic://my.app/search?query=" + query)).start();
  }

  private void onResultsReceived(List<Location> locations) {
    mProgressBar.setVisibility(View.GONE);
    mRecyclerView.setVisibility(View.VISIBLE);
    Adapter adapter = new Adapter(locations);
    mRecyclerView.setAdapter(adapter);
  }

  private void fetchResults(String urlWithQuery) {
    try {
      StringBuilder builder = new StringBuilder();
      URL url = new URL(urlWithQuery);
      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
      int data;
      while ((data = connection.getInputStream().read()) != -1) {
        builder.append((char) data);
      }
      String json = builder.toString();
      List<Location> locations = new Gson().fromJson(json, new TypeToken<List<Location>>(){}.getType());
      onResultsReceived(locations);
    } catch (IOException e) {
      Log.d("MyApp", "Something went wrong when fetching results.  Probably the fake URL! " + e.getMessage());
    }
  }

  private static class Adapter extends RecyclerView.Adapter<Adapter.ViewHolder> {

    private List<Location> mLocations;

    public Adapter(List<Location> locations) {
      mLocations = locations;
    }

    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
      return new ViewHolder(new TextView(parent.getContext()));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
      Location location = mLocations.get(position);
      holder.mTextView.setText(location.getStreetAddress() + ", " + location.getCity() + ", " + location.getCountry());
    }

    @Override
    public int getItemCount() {
      return mLocations.size();
    }

    private static class ViewHolder extends RecyclerView.ViewHolder {
      private TextView mTextView;
      public ViewHolder(View itemView) {
        super(itemView);
        mTextView = (TextView) itemView;
      }
    }
  }
}

File: Android/kotlin/app/src/main/java/org/oreilly/nmd/SearchResultsActivity.kt
package org.oreilly.nmd

import android.app.Activity
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.android.synthetic.main.activity_search.*
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

class SearchResultsActivity : Activity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search)
        search_results_recyclerview.layoutManager = LinearLayoutManager(this)
        val query = intent.getStringExtra(QUERY)
        Thread { fetchResults("http://my.app/search?query=$query") }.start()
    }

    private fun onResultsReceived(locations: List<Location>) {
        search_progress.visibility = View.GONE
        search_results_recyclerview.visibility = View.VISIBLE
        search_results_recyclerview.adapter = Adapter(locations)
    }

    private fun fetchResults(urlWithQuery: String) {
        try {
            val builder = StringBuilder()
            val url = URL(urlWithQuery)
            val connection = url.openConnection() as HttpURLConnection
            var data: Int = connection.inputStream.read()
            while (data != -1) {
                builder.append(data.toChar())
                data = connection.inputStream.read()
            }
            val json = builder.toString()
            val locations = Gson().fromJson<List<Location>>(json, object : TypeToken<List<Location>>() {}.type)
            onResultsReceived(locations)
        } catch (e: IOException) {
            Log.d("MyApp", "Something went wrong when fetching results.  Probably the fake URL! " + e.message)
        }

    }

    private class Adapter(private val locations: List<Location>) : RecyclerView.Adapter<Adapter.ViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(TextView(parent.context))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val (streetAddress, city, country) = locations[position]
            holder.textView.text = "$streetAddress, $city, $country"
        }

        override fun getItemCount(): Int {
            return locations.size
        }

        private class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            val textView: TextView = itemView as TextView
        }
    }

    companion object {
        val QUERY = "QUERY"
    }

}
